src/gui_common/CustomRichTextLabel.cs

```diff
@@ -0,0 +1,245 @@
+                continue;
+            }
+
+            // Character is not a tag, write it normally into the final string
+            if (!isIteratingTag && !isIteratingContent)
+            {
+                result.Append(character);
+            }
+
+            if (isIteratingTag)
+            {
+                // Keep iterating until we hit a closing bracket
+                if (character != ']')
+                {
+                    currentTagBlock.Append(character);
+
+                    // No closing bracket found, just write normally to the final string and abort trying to parse
+                    if (character == '[' || index == ExtendedBbcode.Length - 1)
+                    {
+                        result.Append($"[{currentTagBlock}");
+                        isIteratingTag = false;
+                    }
+
+                    continue;
+                }
+
+                // Closing bracket encountered, proceed with validating the tag
+
+                var tagBlock = currentTagBlock.ToString();
+
+                // Namespace divisor not found
+                if (!tagBlock.Contains(":"))
+                {
+                    result.Append($"[{tagBlock}]");
+                    isIteratingTag = false;
+                    continue;
+                }
+
+                var splitTagBlock = tagBlock.Split(":");
+
+                // Invalid tag syntax, probably not a thrive tag or missing a part
+                if (splitTagBlock.Length != 2)
+                {
+                    result.Append($"[{tagBlock}]");
+                    isIteratingTag = false;
+                    continue;
+                }
+
+                var tagNamespace = splitTagBlock[0];
+                var tagIdentifier = splitTagBlock[1];
+
+                // Not a thrive custom tag, don't parse this
+                if (!tagNamespace.Contains("thrive"))
+                {
+                    result.Append($"[{tagBlock}]");
+                    isIteratingTag = false;
+                    continue;
+                }
+
+                // Tag seems okay, next step is to try parse the tagged substring and closing tag
+
+                // Is a closing tag
+                if (tagNamespace.BeginsWith("/"))
+                {
+                    // Closing tag doesn't match opening tag or vice versa, aborting parsing
+                    if (tagStack.Count == 0 || tagStack.Peek() != tagIdentifier)
+                    {
+                        result.Append($"[{tagBlock}]");
+                        isIteratingTag = false;
+                        continue;
+                    }
+
+                    // Finally try building the bbcode template for the tagged substring
+
+                    var closingTagStartIndex = ExtendedBbcode.Find("[", lastStartingTagEndIndex);
+
+                    var input = ExtendedBbcode.Substr(
+                        lastStartingTagEndIndex + 1, closingTagStartIndex - lastStartingTagEndIndex - 1);
```

Comment:
Like I'd imagine this similar sub string add would be used instead of character by character append to the result.